onnxruntime/core/optimizer/nchwc_transformer.cc

```diff
@@ -0,0 +1,742 @@
+  auto& input_defs = node.MutableInputDefs();
+  auto& output_defs = node.MutableOutputDefs();
+
+  // Require that the weights tensor be static.
+  const ONNX_NAMESPACE::TensorProto* conv_W_tensor_proto = nullptr;
+  if (!graph_utils::NodeArgIsConstant(graph_, *input_defs[1]) ||
+      !graph_.GetInitializedTensor(input_defs[1]->Name(), conv_W_tensor_proto) ||
+      (conv_W_tensor_proto->data_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) ||
+      (conv_W_tensor_proto->dims_size() != 4)) {
+    return;
+  }
+
+  const int64_t output_channels = conv_W_tensor_proto->dims(0);
+  const int64_t input_channels = conv_W_tensor_proto->dims(1);
+
+  int64_t group_count;
+  auto* group_attr = GetAttribute(node, "group");
+  if (group_attr != nullptr && group_attr->has_i()) {
+    group_count = group_attr->i();
+  } else {
+    group_count = 1;
+  }
+
+  const size_t nchwc_block_size = MlasNchwcGetBlockSize();
+
+  const int64_t nchwc_output_channels = (output_channels + nchwc_block_size - 1) & ~(nchwc_block_size - 1);
+
+  bool do_reorder_input = true;
+  bool reorder_filter_OIHWBo = false;
+
+  if (group_count > 1) {
+    if ((output_channels % nchwc_block_size) != 0) {
+      return;
+    }
+    if (input_channels == 1 && output_channels == group_count) {
+      // Depthwise convolution.
+      reorder_filter_OIHWBo = true;
+    } else if (((input_channels % nchwc_block_size) != 0) ||
+               ((output_channels % group_count) != 0) ||
+               (((output_channels / group_count) % nchwc_block_size) != 0)) {
+      return;
+    }
+  } else {
+    if (static_cast<size_t>(input_channels) < nchwc_block_size) {
+      // Use NCHW input buffer directly.
+      reorder_filter_OIHWBo = true;
+      do_reorder_input = false;
+    } else if ((input_channels % nchwc_block_size) != 0) {
+      return;
+    }
+  }
+
+  // Also require that the optional bias tensor be static.
+  const ONNX_NAMESPACE::TensorProto* conv_B_tensor_proto = nullptr;
+  if (input_defs.size() >= 3) {
+    if (!graph_utils::NodeArgIsConstant(graph_, *input_defs[2]) ||
+        !graph_.GetInitializedTensor(input_defs[2]->Name(), conv_B_tensor_proto) ||
+        (conv_B_tensor_proto->data_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) ||
+        (conv_B_tensor_proto->dims_size() != 1) ||
+        (conv_B_tensor_proto->dims(0) != output_channels)) {
+      return;
+    }
+  }
+
+  // Check if the filter has already been converted to the target format.
+  std::unordered_map<NodeArg*, NodeArg*>* filters_map;
+  if (reorder_filter_OIHWBo) {
+    filters_map = &filters_OIHWBo_;
+  } else {
+    filters_map = &filters_OIHWBiBo_;
+  }
+
+  NodeArg* nchwc_conv_W_arg;
+  auto filters_it = filters_map->find(input_defs[1]);
+  if (filters_it != filters_map->end()) {
+    // Reuse the existing NodeArg.
+    nchwc_conv_W_arg = filters_it->second;
+  } else {
+    auto conv_W = std::make_unique<Initializer>(conv_W_tensor_proto);
```

Comment:
Do we need to create a unique pointer here? I think just creating a temp tensor_proto object would suffice (since we are not moving conv_W)?

Reply:
I was following existing practice from other transformers/rules. Can I punt this one for now?

Reply:
There are other changes I would do to initializer to make that lighter weight: the current copy of an element at a time is painful but I didn't want that mixed with this.

Reply:
yeah let's leave it for now -- we need to make it easier to create initializers in the future. I went through that in the ShapeToInitializer rule and since we have it here as well, it is worth creating a helper method to make it easier. But let's do it in another PR.